worker/leadership/tracker.go

```diff
@@ -121,6 +122,75 @@ func (t *Tracker) WaitMinion() leadership.Ticket {
 	return t.submit(t.waitMinionTickets)
 }
 
+// WithoutLeadershipChange is part of the leadership.Tracker interface.
+func (t *Tracker) WithoutLeadershipChange(
+	ctx context.Context, fn func(context.Context) error,
+) error {
+	// Check that the context has not been not cancelled
+	// before we start the operation.
+	if err := ctx.Err(); err != nil {
+		return errors.Annotate(err, "context cancelled before executing leadership func")
+	}
+
+	// leadershipCtx is cancelled if leadership changes.
+	leadershipCtx, leadershipCancel := context.WithCancel(context.Background())
+	defer leadershipCancel()
+
+	// waitCtx ensures the go routine exits even
+	// without a leadership change.
+	waitCtx, waitCancel := context.WithCancel(ctx)
+	defer waitCancel()
+
+	// Start will be closed when we start waiting for leadership to change.
+	start := make(chan struct{})
+
+	// First get the correct channel to wait on, depending if we are the leader or not.
+	var leadershipChanged <-chan struct{}
+	claimLeader := t.ClaimLeader()
+	ready := claimLeader.Ready()
+	select {
+	case <-ctx.Done():
+		return errors.Annotate(ctx.Err(), "context cancelled before executing leadership func")
+	case <-ready:
+		isLeader := claimLeader.Wait()
+		if isLeader {
+			leadershipChanged = t.WaitMinion().Ready()
+		} else {
+			leadershipChanged = t.WaitLeader().Ready()
+		}
+	}
+
+	// Cancel the leadershipCtx when leadership changes.
+	go func() {
+		close(start)
+		select {
+		case <-waitCtx.Done():
+		case <-leadershipChanged:
+			leadershipCancel()
+		}
+	}()
+
+	// Wait for the go routine to start.
+	select {
+	case <-ctx.Done():
+		return errors.Annotate(ctx.Err(), "context cancelled before executing leadership func")
+	case <-leadershipCtx.Done():
+		// Exit early is leadership changes before we even start the operation.
+		return leadership.ErrLeadershipChanged
+	case <-start:
+	}
+
+	// Run the func with the leadershipCtx.
+	err := fn(leadershipCtx)
+	if errors.Is(leadershipCtx.Err(), context.Canceled) {
+		return leadership.ErrLeadershipChanged
+	}
+	if err != nil {
+		return errors.Annotate(err, "executing leadership func")
+	}
+	return ctx.Err()
+}
+
```

Comment:
This can be simplified significantly.

```suggestion
// WithoutLeadershipChange is part of the leadership.Tracker interface.
func (t *Tracker) WithoutLeadershipChange(
	ctx context.Context, fn func(context.Context) error,
) error {
	// Check that the context has not been not cancelled
	// before we start the operation.
	if err := ctx.Err(); err != nil {
		return errors.Annotate(err, "context cancelled before executing leadership func")
	}

	// ctx is cancelled if leadership changes.
	ctx, cancel := context.WithCancelCause(ctx)
	defer cancel(context.Canceled)

	// First get the correct channel to wait on, depending if we are the leader or not.
	var leadershipChanged <-chan struct{}
	claimLeader := t.ClaimLeader()
	ready := claimLeader.Ready()
	select {
	case <-ctx.Done():
		return errors.Annotate(ctx.Err(), "context cancelled before executing leadership func")
	case <-ready:
		isLeader := claimLeader.Wait()
		if isLeader {
			leadershipChanged = t.WaitMinion().Ready()
		} else {
			leadershipChanged = t.WaitLeader().Ready()
		}
	}

	// Cancel the ctx when leadership changes.
	go func() {
		select {
		case <-ctx.Done():
		case <-leadershipChanged:
			cancel(leadership.ErrLeadershipChanged)
		}
	}()

	// Run the func with the ctx.
	err := fn(ctx)
	if errors.Is(err, context.Canceled) {
		return context.Cause(ctx)
	} else if err != nil {
		return errors.Annotate(err, "executing leadership func")
	}
	return ctx.Err()
}

```

Reply:
Using WithCancelCause is a good idea, thanks.